SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularitySlaveAndRackManager.java

```diff
@@ -208,13 +210,23 @@ public SlaveMatchState doesOfferMatch(SingularityOfferHolder offerHolder, Singul
         }
         break;
       case OPTIMISTIC:
-        final double numPerSlave = numDesiredInstances / (double) stateCache.getNumActiveSlaves();
+        Collection<SingularityTaskId> currentlyActiveTasksForRequestClusterwide = stateCache.getActiveTaskIdsForRequest(taskRequest.getRequest().getId());
 
-        final boolean isSlaveOk = numOnSlave < numPerSlave;
+        // If no tasks are active for this request yet, we can fall back to greedy.
+        if (currentlyActiveTasksForRequestClusterwide.size() > 0) {
```

Comment:
Hmm, so the general pattern in this method is to return only if the `SlaveMatchState` is determined to be something other than `OK`. Otherwise, flow is allowed to trickle down to that final `return SlaveMatchState.OK` statement.